Advancement datagen problem related to parent

I get this error when I try to do the advancement datagen.
In 1.20.1 it works and I don't have this kind of parent related error.
https://github.com/railcraft-reborn/railcraft/blob/f162ad97205e17423013a4918d07038de9543553/src/main/java/mods/railcraft/data/advancements/RailcraftTrackAdvancements.java#L50


```
[15:08:28] [main/INFO] [minecraft/DataGenerator]: Starting provider: Advancements
Exception in thread "main" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:32)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.Launcher.run(Launcher.java:108)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.Launcher.main(Launcher.java:78)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23)
	at cpw.mods.bootstraplauncher@1.1.2/cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:141)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at MC-BOOTSTRAP/fml_loader@1.0.9/net.neoforged.fml.loading.targets.CommonLaunchHandler.runTarget(CommonLaunchHandler.java:126)
	at MC-BOOTSTRAP/fml_loader@1.0.9/net.neoforged.fml.loading.targets.CommonLaunchHandler.dataService(CommonLaunchHandler.java:122)
	at MC-BOOTSTRAP/fml_loader@1.0.9/net.neoforged.fml.loading.targets.ForgeDataUserdevLaunchHandler.runService(ForgeDataUserdevLaunchHandler.java:22)
	at MC-BOOTSTRAP/fml_loader@1.0.9/net.neoforged.fml.loading.targets.CommonLaunchHandler.lambda$launchService$4(CommonLaunchHandler.java:108)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:30)
	... 7 more
Caused by: java.util.concurrent.CompletionException: java.lang.IllegalStateException: The parent: 'railcraft:tracks/root' of advancement 'railcraft:tracks/firestone', has not been saved yet!
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1194)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
	at TRANSFORMER/minecraft@1.20.2/net.minecraft.data.advancements.AdvancementProvider.run(AdvancementProvider.java:35)
	at TRANSFORMER/minecraft@1.20.2/net.minecraft.data.HashCache.generateUpdate(HashCache.java:94)
	at TRANSFORMER/minecraft@1.20.2/net.minecraft.data.DataGenerator.lambda$run$0(DataGenerator.java:44)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
	at TRANSFORMER/minecraft@1.20.2/net.minecraft.data.DataGenerator.run(DataGenerator.java:37)
	at TRANSFORMER/neoforge@20.2.21-beta/net.neoforged.neoforge.data.event.GatherDataEvent$DataGeneratorConfig.lambda$runAll$2(GatherDataEvent.java:134)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.9/cpw.mods.modlauncher.api.LamdbaExceptionUtils.lambda$rethrowConsumer$0(LamdbaExceptionUtils.java:34)
	at java.base/java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:647)
	at TRANSFORMER/neoforge@20.2.21-beta/net.neoforged.neoforge.data.event.GatherDataEvent$DataGeneratorConfig.runAll(GatherDataEvent.java:130)
	at TRANSFORMER/neoforge@20.2.21-beta/net.neoforged.neoforge.data.loading.DatagenModLoader.begin(DatagenModLoader.java:53)
	at TRANSFORMER/minecraft@1.20.2/net.minecraft.data.Main.main(Main.java:92)
	... 16 more
Caused by: java.lang.IllegalStateException: The parent: 'railcraft:tracks/root' of advancement 'railcraft:tracks/firestone', has not been saved yet!
	at TRANSFORMER/neoforge@20.2.21-beta/net.neoforged.neoforge.common.extensions.IAdvancementBuilderExtension.save(IAdvancementBuilderExtension.java:36)
	at TRANSFORMER/railcraft@1.0.10/mods.railcraft.data.advancements.RailcraftTrackAdvancements.generate(RailcraftTrackAdvancements.java:50)
	at TRANSFORMER/neoforge@20.2.21-beta/net.neoforged.neoforge.common.data.AdvancementProvider$AdvancementGenerator.lambda$toSubProvider$0(AdvancementProvider.java:60)
	at TRANSFORMER/minecraft@1.20.2/net.minecraft.data.advancements.AdvancementProvider.lambda$run$2(AdvancementProvider.java:48)
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
	... 28 more

Process finished with exit code 1
```

--- a/src/main/java/net/neoforged/neoforge/common/extensions/IAdvancementBuilderExtension.java
+++ b/src/main/java/net/neoforged/neoforge/common/extensions/IAdvancementBuilderExtension.java
@@ -39,6 +39,7 @@ public interface IAdvancementBuilderExtension {
         }
 
         saver.accept(advancementholder);
+        fileHelper.trackGenerated(id, PackType.SERVER_DATA, ".json", "advancements");
         return advancementholder;
     }
 }

--- a/tests/src/generated/resources/data/minecraft/advancements/obtain_diamond_block.json
+++ b/tests/src/generated/resources/data/minecraft/advancements/obtain_diamond_block.json
@@ -1,4 +1,5 @@
 {
+  "parent": "data_gen_test:obtain_dirt",
   "criteria": {
     "obtained_diamond_block": {
       "conditions": {

--- a/tests/src/main/java/net/neoforged/neoforge/debug/DataGeneratorTest.java
+++ b/tests/src/main/java/net/neoforged/neoforge/debug/DataGeneratorTest.java
@@ -912,25 +912,28 @@ public class DataGeneratorTest {
 
         @Override
         public void generate(HolderLookup.Provider registries, Consumer<AdvancementHolder> saver, ExistingFileHelper existingFileHelper) {
-            Advancement.Builder.advancement().display(Items.DIRT,
-                    Component.translatable(Items.DIRT.getDescriptionId()),
-                    Component.translatable("dirt_description"),
-                    new ResourceLocation("textures/gui/advancements/backgrounds/stone.png"),
-                    FrameType.TASK,
-                    true,
-                    true,
-                    false)
+            var obtainDirt = Advancement.Builder.advancement()
+                    .display(Items.DIRT,
+                            Component.translatable(Items.DIRT.getDescriptionId()),
+                            Component.translatable("dirt_description"),
+                            new ResourceLocation("textures/gui/advancements/backgrounds/stone.png"),
+                            FrameType.TASK,
+                            true,
+                            true,
+                            false)
                     .addCriterion("has_dirt", InventoryChangeTrigger.TriggerInstance.hasItems(Items.DIRT))
                     .save(saver, new ResourceLocation(MODID, "obtain_dirt"), existingFileHelper);
 
-            Advancement.Builder.advancement().display(Items.DIAMOND_BLOCK,
-                    Component.translatable(Items.DIAMOND_BLOCK.getDescriptionId()),
-                    Component.literal("You obtained a DiamondBlock"),
-                    new ResourceLocation("textures/gui/advancements/backgrounds/stone.png"),
-                    FrameType.CHALLENGE,
-                    true,
-                    true,
-                    false)
+            Advancement.Builder.advancement()
+                    .parent(obtainDirt)
+                    .display(Items.DIAMOND_BLOCK,
+                            Component.translatable(Items.DIAMOND_BLOCK.getDescriptionId()),
+                            Component.literal("You obtained a DiamondBlock"),
+                            new ResourceLocation("textures/gui/advancements/backgrounds/stone.png"),
+                            FrameType.CHALLENGE,
+                            true,
+                            true,
+                            false)
                     .addCriterion("obtained_diamond_block", InventoryChangeTrigger.TriggerInstance.hasItems(Items.DIAMOND_BLOCK))
                     .save(saver, new ResourceLocation("obtain_diamond_block"), existingFileHelper);
 

